include/drivers/video.h

```diff
@@ -0,0 +1,301 @@
+/**
+ * @file
+ *
+ * @brief Public APIs for Video.
+ */
+
+/*
+ * Copyright (c) 2019 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_VIDEO_H_
+#define ZEPHYR_INCLUDE_VIDEO_H_
+
+/**
+ * @brief Video Interface
+ * @defgroup video_interface Video Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <device.h>
+#include <stddef.h>
+#include <zephyr.h>
+
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+#define VIDEO_CAP_CAPTURE	0x00000001 /* Video capture device/buffer */
+#define VIDEO_CAP_OUTPUT	0x00000002 /* Video output device/buffer */
+#define VIDEO_CAP_M2M		0x00000004 /* Video memory to memory device */
+
+/**
+ * @brief video format structure
+ *
+ * Used to configure frame format.
+ *
+ * @param pixelformat is the fourcc pixel format value.
+ * @param width is the frame width in pixels.
+ * @param height is the frame height in pixels.
+ * @param pitch is the line stride, the number of bytes that needs to be added
+ *    to the address in the first pixel of a row in order to go to the address
+ *    of the first pixel of the next row (>=width).
+ */
+struct video_format {
+	u32_t pixelformat;
+	u32_t width;
+	u32_t height;
+	u32_t pitch;
+};
+
+/**
+ * @brief video buffer structure
+ *
+ * Represent a video frame.
+ *
+ * @param driver_data is a pointer to driver specific data.
+ * @param frame is a pointer to the frame buffer containing pixels.
+ * @param bytesused is the number of bytes occupied by the data in the buffer.
+ * @param type is the type of buffer (capture, output).
+ * @param fmt is the video frame format.
+ */
+struct video_buffer {
+	void *driver_data;
+	void *frame;
+	u32_t bytesused;
+	u32_t type;
```

Comment:
I think that instead of different types of buffer (which might be CAPTURE for one component and OUTPUT for other) we should have multiple endpoints in the video devices. For in & out, there will be usually one endpoint, for M2M, two.
So you will put filled buffer to "IN" and empty to "OUT" in order to process data.

Please note that this way we can represent also things like mixers and muxers.

Reply:
Good idea, it makes sense indeed and is much more scalable.

Reply:
I pushed a new version and added endpoint support with basic IN and OUT eps for now.

Reply:
Now the API looks better.

I apologize, but in next 3 weeks I will be not able to assist you with this PR. I strongly recommend you to look into existing video APIs (for example OpenMAX and Gstreramer), as this might help you design simple and easy to use interface. For example, the endpoint idea comes directly from these frameworks.

Reply:
Thank you for your support/ideas, I don't think there is a hurry, I'm going to look closer at OpenMAX and gstreamer designs. Note that the original version of this API is already inspired by the V4L2/Media Linux subsystem.

Reply:
Good. The more different API you see, the better one you create :).
I personally have to look into V4L2, as so far i worked a bit higher.